### PR TITLE
Add popovers to show glossary entries in-page

### LINKED ISSUE #70: Show glossary definitions in-text
Allow users to view definitions of terms in the glossary without leaving the page. I'm picturing something like Wikipedia's preview cards:

Screen Shot 2022-01-31 at 11 43 10 AM

### PR DESCRIPTION
**Pull request by molly**
This moves all glossary entries from being hardcoded in a JSX file to being stored in the DB, and adds support for inline definitions to open in a popover style. All definitions need to be explicitly marked in the entry body, like so: `<button class="define-target" id="stablecoin">stablecoin</button>`. The class must be `define-target` and the id must specifically match one of the glossary entry IDs. The text within the button can be anything.

Terms marked as such will render like so:
Screen Shot 2022-02-04 at 9 10 36 PM

And display a popup on click:
Screen Shot 2022-02-04 at 9 10 57 PM

<details>
<summary>Original PR description (retained to provide context to below conversation)</summary>
Putting up a WIP in case there are people more familiar with popover positioning than I am who can help me out.

I've got [react-tiny-popover](https://github.com/alexkatz/react-tiny-popover) rendering some nice popovers, but the positioning is whacky. I'm setting as the target the `<span>` containing the underlined word to define, but the popover is showing up in somewhat weird places. I've tried manually setting the position to the location of the span or even the click target but that just made the position wrong in new and exciting ways.

I've put this on a [dev server](https://web3-dev-340101.uc.r.appspot.com/) if anyone wants to poke at a live version. To repro locally, clone the project, `cd` into the `src` directory and then `npm install` and `npm run dev`. For both live and local versions you'll need to sort the timeline by "ascending" (I have only added anchors to a few of the oldest entries). Clicking on the underlined terms should open the popover.

## Not bad...
Screen Shot 2022-02-01 at 7 54 35 PM

## Not even a little close
Feb-01-2022 19-58-57

If I can fix the positioning problem, this will close #70.
</details>

### CODE REVIEW

File: src/components/timeline/EnrichedBody.js [hunk 1]
```diff
@@ -0,0 +1,58 @@
...
+    );
+  };
+
+  return (
+    <Popover
+      content={renderPopoverContents}
+      isOpen={!!activeTarget}
+      onClickOutside={() => setActiveTarget(null)}
+      containerClassName="definition-popover"
+    >
```
[the remark this answers is not in the document]
Comment (kienankb) [reply src/components/timeline/EnrichedBody.js:45]: seems like setting either `boundaryElement` or `parentElement` to `activeTarget` will do this, too

**Comment by rlahoda** [comment]
Looking this, I think the problem lies in where the popover is located. `position: absolute` relies on the closest parent that is positioned. So looking at `EnrichedBody.js` it looks like the popover parent is the `body` tag. I'm not familiar enough with how you have it set up but I would suggest that if there's a way to have the `<Popover>` component added into each individual `<Entry>`, then when it is loaded, the parent can be the `<span>` with `position: relative` set to it and then `position: absolute` will behave. I will see if I can figure it out a bit more and get it working, but hopefully that might be helpful.

**Comment by kienankb** [comment]
in combination with what @eatspaint found--maybe it has something to do with the fact that the `Popup`'s child content is the whole body, and not just the definable term's `span` (even though that `span` _does_ end up being the `activeTarget`, which is confusing)?

**Comment by eatspaint** [comment]
Probably unrelated to this issue, but it may cause headache in the future: it looks like glossary terms are being applied as the `id` attribute on those `define-term` buttons. If a term shows up more than once on the page, things might get weird from the duped `id`s

**Comment by molly** [comment]
@rlahoda Despite the name, `EnrichedBody` is rendered as a child of the `Entry` component and so the Popover should be a direct descendent of the relatively-positioned `<div className="timeline-description">` already 😞 

> in combination with what eatspaint found--maybe it has something to do with the fact that the `Popup`'s child content is the whole body, and not just the definable term's `span` (even though that `span` _does_ end up being the `activeTarget`, which is confusing)?

@kienankb The entry body (`span`s and all) is coming from the DB, and I'm hoping I'll be able to just propagate the events up through the container `span` rather than having to try to interleave React into the HTML. It's a little janky, but I went down the path of trying to render React elements into the stored HTML and that felt way jankier.

> Probably unrelated to this issue, but it may cause headache in the future: it looks like glossary terms are being applied as the `id` attribute on those `define-term` buttons. If a term shows up more than once on the page, things might get weird from the duped `id`s

@eatspaint Oh good call.

**Comment by dceddia** [comment]
The popover is setting its own `translate` that seems to be throwing it way off. It [looks like](https://github.com/alexkatz/react-tiny-popover/blob/1fb6c067c90815afa8a07f5ace88a04a636dc924/src/usePopover.ts#L65-L70) that value is based on the parent element's position. Setting that seems to help somewhat, but I think I've got the wrong element as the parent. It's positioned on the parent div. I think if the `parentElement` was the actual underlined link that might do it.

image


```
export default function EnrichedBody({ children, glossary }) {
  const [activeTarget, setActiveTarget] = useState(null);
  const self = useRef();

  /* more stuff */

  return (
    <div ref={self}>
      <Popover
        parentElement={self.current}
        content={renderPopoverContents}
        isOpen={!!activeTarget}
        onClickOutside={() => setActiveTarget(null)}
        containerClassName="definition-popover"
      >
        <span
          id="bodyContainer"
          dangerouslySetInnerHTML={{ __html: children }}
          onClick={callback}
        />
      </Popover>
    </div>
  );
}
```

On a different note, I tried setting the `contentLocation` prop to a function, to see if I could figure out how it was deciding where to put the thing, and omg it [calls that on EVERY FRAME](https://github.com/alexkatz/react-tiny-popover/blob/master/src/Popover.tsx#L88-L130) while the popup is open and uses like 10% CPU 😂

**Comment by rlahoda** [comment]
When I inspected the popover and entry, they're not rendering in the html with the popover as a child of the element, the popover is showing up as a child of the body
Here's the entry showing its ancestry
Screen Shot 2022-02-01 at 9 45 11 PM
Here's the popover showing its ancestry
Screen Shot 2022-02-01 at 9 48 51 PM
So I'm wondering if maybe the popover package itself is the problem? I noticed there's a `<div class="react-tiny-popover-scout" ` right below that so perhaps the package is adding in the popovers outside of where you were expecting them to go?

**Comment by dceddia** [comment]
Oh! Even simpler, skip adding the `div` around the Popover like I did, and just set `parentElement={activeTarget}`. But... even with that it doesn't work right 🤔 The fact that it's perfectly positioned over *stablecoin* seems to be a happy accident, because on the other ones it's in the same spot now: always some percentage from the left edge of the text block.

**Comment by molly** [comment]
@rlahoda I think that's a pretty common property of popover libraries. They just smack it on to the end of the body (regardless of where it's actually located in the DOM) and then position it after the fact.

**Comment by molly** [comment]
I think I've got enough to go on here, though, now that it's properly attaching to the entry text!

**Comment by dceddia** [comment]
I think I've got it! 🎉  Adding the `contentLocation` gets it into place above the link. And it didn't like when `parentElement` was null.

```
      parentElement={activeTarget || document.body}
      contentLocation={(s) => {
        return {
          top: -s.popoverRect.height - 10,
          left: 0
        }
      }}
```

I was wrong when I thought it was based on the parent's position. It looks like it's actually based on the *child element's position* – the actual entry text block. So you could probably wrap the actual underlined `<button>` with the `<Popover>` too.

**Pull request merged**